packages/jupyter-ai/jupyter_ai/extension.py

```diff
@@ -238,41 +232,72 @@ class AiExtension(ExtensionApp):
+        """
+
+        self.ychats_by_room: Dict[str, YChat] = {}
+        """Cache of YChat instances, indexed by room ID."""
+
         self.event_logger = self.serverapp.web_app.settings["event_logger"]
         self.event_logger.add_listener(
             schema_id=JUPYTER_COLLABORATION_EVENTS_URI, listener=self.connect_chat
         )
 
-        # Keep the message indexes to avoid extra computation looking for a message when
-        # updating it.
-        self.messages_indexes = {}
-
     async def connect_chat(
         self, logger: EventLogger, schema_id: str, data: dict
     ) -> None:
-        if (
+        # ignore events that are not chat room initialization events
+        if not (
             data["room"].startswith("text:chat:")
             and data["action"] == "initialize"
             and data["msg"] == "Room initialized"
         ):
+            return
 
-            self.log.info(f"Collaborative chat server is listening for {data['room']}")
-            chat = await self.get_chat(data["room"])
+        # log room ID
+        room_id = data["room"]
+        self.log.info(f"Connecting to a chat room with room ID: {room_id}.")
 
-            if chat is None:
-                return
+        # get YChat document associated with the room
+        ychat = await self.get_chat(data["room"])
+        if ychat is None:
+            return
 
-            # Add the bot user to the chat document awareness.
-            BOT["avatar_url"] = url_path_join(
-                self.settings.get("base_url", "/"), "api/ai/static/jupyternaut.svg"
-            )
-            if chat.awareness is not None:
-                chat.awareness.set_local_state_field("user", BOT)
+        # Add the bot user to the chat document awareness.
+        BOT["avatar_url"] = url_path_join(
+            self.settings.get("base_url", "/"), "api/ai/static/jupyternaut.svg"
+        )
+        if ychat.awareness is not None:
+            ychat.awareness.set_local_state_field("user", BOT)
+
+        # initialize chat handlers for new chat
+        self.chat_handlers_by_room[room_id] = self._init_chat_handlers(ychat)
 
-            callback = partial(self.on_change, chat)
-            chat.ymessages.observe(callback)
+        callback = partial(self.on_change, room_id)
+        ychat.ymessages.observe(callback)
 
     async def get_chat(self, room_id: str) -> Optional[YChat]:
+        """
+        Retrieves the YChat instance associated with a room ID. This method
+        is cached, i.e. successive calls with the same room ID quickly return a
+        cached value.
+
+        TODO: Determine if get_chat() should ever fail under normal usage
+        scenarios. If not, we should just raise an exception if chat is `None`,
+        and indicate the return type as just `YChat` instead of
+        `Optional[YChat]`. This will simplify the code by removing redundant
+        null checks.
+        """
+        if room_id in self.ychats_by_room:
```

Comment:
See comment above about ychats_by_room, but AFAIU this assessment should be always False.